src/analysis/useless_initializer.d

```diff
@@ -25,45 +32,128 @@ final class UselessInitializerChecker : BaseAnalyzer
+            return;
+
+        _structStack.insert(decl.name.text);
+        _structCanBeInit[decl.name.text] = false;
+        _atDisabled.insert(false);
+        decl.accept(this);
+        _structStack.removeBack();
+        _atDisabled.removeBack();
+    }
+
+    override void visit(const(Declaration) decl)
+    {
+        _inStruct.insert(decl.structDeclaration !is null);
+        decl.accept(this);
+        if (_inStruct.length > 1 && _inStruct[$-2] && decl.constructor &&
+            ((decl.constructor.parameters && decl.constructor.parameters.parameters.length == 0) ||
+            !decl.constructor.parameters))
+        {
+            _atDisabled[$-1] = !decl.attributes
+                .filter!(a => a.atAttribute !is null && a.atAttribute.identifier.text == "disable")
+                .empty;
+        }
+        _inStruct.removeBack();
+    }
+
+    override void visit(const(Constructor) decl)
+    {
+        if (_inStruct.length > 1 && _inStruct[$-2] &&
+            ((decl.parameters && decl.parameters.parameters.length == 0) || !decl.parameters))
+        {
+            _structCanBeInit[_structStack.back()] = !_atDisabled[$-1];
+            if (!_structCanBeInit[_structStack.back()])
+                _structCanBeInit[_structStack.back()] = !decl.memberFunctionAttributes
+                    .filter!(a => a.atAttribute !is null && a.atAttribute.identifier.text == "disable")
+                    .empty;
+        }
+        decl.accept(this);
+    }
+
+    // issue 473, prevent to visit delegates that contain duck type checkers.
+    override void visit(const(TypeofExpression)) {}
+
+    // issue 473, prevent to check expressions in __traits(compiles, ...)
+    override void visit(const(TraitsExpression) e)
+    {
+        if (e.identifier.text == "compiles")
+        {
+            return;
+        }
+        else
+        {
+            e.accept(this);
+        }
     }
 
     override void visit(const(VariableDeclaration) decl)
     {
+        if (!decl.type || !decl.type.type2 ||
+            // initializer has to appear clearly in generated ddoc
+            decl.comment !is null ||
+            // issue 474, manifest constants HAVE to be initialized.
+            !decl.storageClasses.filter!(a => a.token == tok!"enum").empty)
+                return;
+
         foreach (declarator; decl.declarators)
         {
-            if (!decl.type || !decl.type.type2)
-                continue;
-            if (!declarator.initializer || !declarator.initializer.nonVoidInitializer)
-                continue;
-
-            import std.format : format;
+            if (!declarator.initializer || !declarator.initializer.nonVoidInitializer ||
+                declarator.comment !is null)
+                    continue;
 
             version(unittest)
+            {
                 enum warn = q{addErrorMessage(declarator.name.line, declarator.name.column,
                     key, msg);};
+            }
             else
```

Comment:
Off-topic: I am not a huge friend of this pattern. While I do the see merits of being able to just use `X`, the string formatting could (in theory of course) be broken.